File: app/src/main/java/com/example/testproguard/test6/base/module/BaseModule.java
package com.example.testproguard.test6.base.module;

import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.example.testproguard.test6.base.LazyLoader;
import com.example.testproguard.test6.base.presenter.BaseMultiPartMvpPresenter;
import com.example.testproguard.test6.base.view.IBaseMvpView;
import com.example.testproguard.test6.base.view.IBaseTargetView;

/**
 * 模块公共父类
 *
 */
public abstract class BaseModule<V extends IBaseTargetView, P extends BaseMultiPartMvpPresenter> extends LazyLoader implements IBaseMvpView, View.OnClickListener {

    protected Context mContext;
    protected V mTarget;
    protected P mPresenter;

    protected View mCurrentLayout;
    protected View mRootLayout;

    public BaseModule(View parent) {
        super();
        this.mRootLayout = parent;
    }

    public BaseModule(V target, P presenter, View parent) {
        this(parent);
        this.mTarget = target;
        this.mTarget.addModule(this);
        this.mPresenter = presenter;
    }

    @Override
    public void showToast(String str) {
        if(context() != null)
            Toast.makeText(context(), str, Toast.LENGTH_SHORT).show();
    }

    @Override
    protected void lazyInitView() {

    }

    public boolean isOnShowing() {
        if (mCurrentLayout == null)
            return false;
        return mCurrentLayout.getVisibility() == View.VISIBLE;
    }

    public Context getContext() {
        return null;
    }

    public Context getApplicationContext() {
        return null;
    }

    public void setVisible(@Constants.AnimEnum int animEnum) {
        if (animEnum == Constants.MODULE_VISIABLE)
            init();
        if (!isLazyInit() && animEnum == Constants.MODULE_IN_VISIABLE)
            return;

        if (mRootLayout != null)
            mRootLayout.clearAnimation();

        switch (animEnum) {
            case Constants.MODULE_VISIABLE:       // 显示

                break;
            case Constants.MODULE_IN_VISIABLE:    // 隐藏

                break;
        }
    }

    /**********************************************************************************************
     * Activity生命周期
     **********************************************************************************************/
    public void onCreate(Bundle savedInstanceState) {

    }

    public void onStart() {
    }

    public void onRestart() {
    }

    public void onResume() {
    }

    public void onPause() {
    }

    public void onStop() {
    }

    public void onDestroy() {
    }

    /**********************************************************************************************
     * Fragment 生命周期
     **********************************************************************************************/
    public void onAttach(@NonNull Context context) {

    }

    public void onActivityCreated(@Nullable Bundle savedInstanceState) {

    }

    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {

    }

    public void onDestroyView() {

    }

    public void onDetach() {

    }

    @Override
    public void onClick(View v) {

    }

    private Context context() {
        return mContext;
    }
}

File: app/src/main/java/com/example/testproguard/test6/base/module/BaseModule.kt
package com.example.testproguard.test6.base.module

import android.content.Context
import android.os.Bundle
import android.view.View
import android.widget.Toast
import com.example.testproguard.test6.base.LazyLoader
import com.example.testproguard.test6.base.module.Constants.AnimEnum
import com.example.testproguard.test6.base.presenter.BaseMultiPartMvpPresenter
import com.example.testproguard.test6.base.view.IBaseMvpView
import com.example.testproguard.test6.base.view.IBaseTargetView

/**
 * 模块公共父类
 *
 */
abstract class BaseModule<V : IBaseTargetView?, P : BaseMultiPartMvpPresenter<*>?>(@JvmField protected var mRootLayout: View?) :
    LazyLoader(), IBaseMvpView, View.OnClickListener {
    protected var mContext: Context? = null
    protected var mTarget: V? = null
    @JvmField
    protected var mPresenter: P? = null

    @JvmField
    protected var mCurrentLayout: View? = null

    constructor(target: V, presenter: P, parent: View?) : this(parent) {
        this.mTarget = target
        mTarget!!.addModule(this)
        this.mPresenter = presenter
    }

    override fun showToast(str: String) {
        if (context() != null) Toast.makeText(context(), str, Toast.LENGTH_SHORT).show()
    }

    override fun lazyInitView() {
    }

    val isOnShowing: Boolean
        get() {
            if (mCurrentLayout == null) return false
            return mCurrentLayout!!.visibility == View.VISIBLE
        }

    override fun getContext(): Context? {
        return null
    }

    val applicationContext: Context?
        get() = null

    open fun setVisible(@AnimEnum animEnum: Int) {
        if (animEnum == Constants.MODULE_VISIABLE) init()
        if (!isLazyInit && animEnum == Constants.MODULE_IN_VISIABLE) return

        if (mRootLayout != null) mRootLayout!!.clearAnimation()

        when (animEnum) {
            Constants.MODULE_VISIABLE -> {}
            Constants.MODULE_IN_VISIABLE -> {}
        }
    }

    /**********************************************************************************************
     * Activity生命周期
     */
    fun onCreate(savedInstanceState: Bundle?) {
    }

    fun onStart() {
    }

    fun onRestart() {
    }

    fun onResume() {
    }

    fun onPause() {
    }

    fun onStop() {
    }

    fun onDestroy() {
    }

    /**********************************************************************************************
     * Fragment 生命周期
     */
    fun onAttach(context: Context) {
    }

    fun onActivityCreated(savedInstanceState: Bundle?) {
    }

    fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    }

    fun onDestroyView() {
    }

    fun onDetach() {
    }

    override fun onClick(v: View) {
    }

    private fun context(): Context? {
        return mContext
    }
}
